# Practical - 06

#### Step 1: Importing the Libraries #As always, the first step will always include importing the libraries which are the NumPy, Pandas and the Matplotlib

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#### Step 2: Importing the dataset In this step, we shall import the Iris Flower dataset which is stored as Iris.csv and save it to the variable dataset.

In [3]:
dataset = pd.read_csv(r'C:\Users\hasna\Desktop\Simran\TE\TE-SEM-2\Practicals\DSBDL\Datasets\Iris.csv')

In [4]:
dataset.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [5]:
dataset.shape

(150, 6)

In [6]:
gk=dataset.groupby('Species')

In [7]:
gk.first()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
Species                                                                       
Iris-setosa        1            5.1           3.5            1.4           0.2
Iris-versicolor   51            7.0           3.2            4.7           1.4
Iris-virginica   101            6.3           3.3            6.0           2.5

#### Step 3: After this, we assign the 4 independent variables to X and the dependent variable ‘species’ to Y. The first 5 rows of the dataset are displayed.


In [34]:
# iloc[:,:4]: This is a method used to select rows and columns from the DataFrame. `iloc` is used for integer-based indexing, where :,:4 selects all rows (:) and the first four columns 

#before comma it's rows range,after comma it's column range

# .values: This converts the selected data (X) into a NumPy array


X = dataset.iloc[:,:4].values
y = dataset['Species'].values

In [30]:
# print(X)
# print(y)

#### Step 4: Splitting the dataset into the Training set and Test set Once we have obtained our data set, we have to split the data into the training set and the test set. In this dataset, there are 150 rows with 50 rows of each of the 3 classes. As each class is given in a continuous order, we need to randomly split the dataset. Here, we have the test_size=0.2, which means that 20% of the dataset will be used for testing purpose as the test set and the remaining 80% will be used as the training set for training the Naive Bayes classification model.

In [32]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

#### Step 5: Feature Scaling The dataset is scaled down to a smaller range using the Feature Scaling option. In this, both the X_train and X_test values are scaled down to smaller values to improve the speed of the program.

In [37]:
# The StandardScaler is used for standardizing features by removing the mean and scaling to unit variance. 

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

#### Step 6: Training the Naive Bayes Classification model on the Training Set In this step, we introduce the class GaussianNB that is used from the sklearn.naive_bayes library. Here, we have used a Gaussian model, there are several other models such as Bernoulli, Categorical and Multinomial. Here, we assign the GaussianNB class to the variable classifier and fit the X_train and y_train values to it for training purpose.

In [38]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

GaussianNB()

**Bayes' Theorem:** Bayes' theorem calculates the probability of a hypothesis (H) given the evidence (E) using conditional probabilities:

```
 𝑃(𝐻/𝐸)=𝑃(𝐸/𝐻)⋅𝑃(𝐻/𝑃(𝐸)
```
* P(H∣E) is the probability of the hypothesis given the evidence (posterior probability).

* P(E∣H) is the probability of the evidence given the hypothesis (likelihood).

* P(H) is the probability of the hypothesis (prior probability).

* P(E) is the probability of the evidence (normalizing constant).

#### Step 7: Predicting the Test set results Once the model is trained, we use the the classifier.predict() to predict the values for the Test set and the values predicted are stored to the variable y_pred.

In [39]:
y_pred = classifier.predict(X_test)
y_pred

array(['Iris-versicolor', 'Iris-setosa', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-virginica', 'Iris-virginica',
       'Iris-setosa', 'Iris-virginica', 'Iris-setosa', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-versicolor', 'Iris-setosa',
       'Iris-virginica', 'Iris-virginica', 'Iris-virginica',
       'Iris-virginica', 'Iris-virginica', 'Iris-virginica',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-versicolor',
       'Iris-setosa', 'Iris-versicolor', 'Iris-setosa', 'Iris-virginica',
       'Iris-virginica', 'Iris-versicolor', 'Iris-setosa'], dtype='<U15')

#### Step 8: Confusion Matrix and Accuracy This is a step that is mostly used in classification techniques. In this, we see the Accuracy of the trained model and plot the confusion matrix. The confusion matrix is a table that is used to show the number of correct and incorrect predictions on a classification problem when the real values of the Test Set are known. It is of the format

In [40]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)


In [41]:
print('Accuracy: ', accuracy_score(y_test, y_pred))

Accuracy:  1.0


In [42]:
print(cm)

[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


#### Step 9: Comparing the Real Values with Predicted Values In this step, a Pandas DataFrame is created to compare the classified values of both the original Test set (y_test) and the predicted results (y_pred).

In [43]:
df = pd.DataFrame({'Real Values':y_test, 'Predicted Values':y_pred})
df

Real Values Predicted Values
0   Iris-versicolor  Iris-versicolor
1       Iris-setosa      Iris-setosa
2   Iris-versicolor  Iris-versicolor
3   Iris-versicolor  Iris-versicolor
4    Iris-virginica   Iris-virginica
5    Iris-virginica   Iris-virginica
6       Iris-setosa      Iris-setosa
7    Iris-virginica   Iris-virginica
8       Iris-setosa      Iris-setosa
9   Iris-versicolor  Iris-versicolor
10  Iris-versicolor  Iris-versicolor
11  Iris-versicolor  Iris-versicolor
12      Iris-setosa      Iris-setosa
13   Iris-virginica   Iris-virginica
14   Iris-virginica   Iris-virginica
15   Iris-virginica   Iris-virginica
16   Iris-virginica   Iris-virginica
17   Iris-virginica   Iris-virginica
18   Iris-virginica   Iris-virginica
19      Iris-setosa      Iris-setosa
20      Iris-setosa      Iris-setosa
21      Iris-setosa      Iris-setosa
22  Iris-versicolor  Iris-versicolor
23      Iris-setosa      Iris-setosa
24  Iris-versicolor  Iris-versicolor
25      Iris-setosa      Iris-setosa
26   Iris-virginica   Iris-virginica
27   Iris-virginica   Iris-virginica
28  Iris-versicolor  Iris-versicolor
29      Iris-setosa      Iris-setosa